scripts/keyless/keylessDeploy.ts

```diff
@@ -0,0 +1,293 @@
+  const deployData = hardhatEthers.utils.solidityPack(["bytes32", "bytes", "bytes"], [saltHash, bytecode, args]);
+
+  const initBytecode = hardhatEthers.utils.solidityPack(["bytes", "bytes"], [bytecode, args]);
+  const deployInfoPacked = hardhatEthers.utils.solidityPack(
+    ["bytes1", "address", "bytes32", "bytes32"],
+    ["0xff", commonFactory, saltHash, hardhatEthers.utils.solidityKeccak256(["bytes"], [initBytecode])],
+  );
+  const addr = hardhatEthers.utils.solidityKeccak256(["bytes"], [deployInfoPacked]);
+
+  return { predictedAddress: `0x${addr.slice(26)}`, deployData };
+}
+
+async function deployInfraKeyless() {
+  const [signer] = await hardhatEthers.getSigners();
+
+  // create2 factory
+  await deployCommonFactory(signer);
+
+  for (let txInfo of Object.values(infraContracts) as InfraTxInfo[]) {
+    const code = await hardhatEthers.provider.getCode(txInfo.predictedAddress);
+    if (code === "0x") {
+      let fundAddr = {
+        to: txInfo.from,
+        value: hardhatEthers.utils.parseEther("1"),
+      };
+      await signer.sendTransaction(fundAddr);
+
+      await (await hardhatEthers.provider.sendTransaction(txInfo.tx)).wait();
+    }
+  }
+}
+
+async function deployInfraWithSigner() {
+  const [signer] = await hardhatEthers.getSigners();
+
+  // create2 factory
+  await deployCommonFactory(signer);
+
+  for (let txInfo of Object.values(infraContracts) as InfraTxInfo[]) {
+    const block = await hardhatEthers.provider.getBlock("latest");
+    const baseBlockFee =
+      block && block.baseFeePerGas ? block.baseFeePerGas : hardhatEthers.utils.parseUnits("1", "gwei");
+    const feeData = await hardhatEthers.provider.getFeeData();
+    const baseMaxFeePerGas = baseBlockFee.mul(2);
+    const maxPriorityFeePerGas = BigNumber.from(feeData.maxPriorityFeePerGas);
+
+    // get init bytecode
+    const deployData = await hardhatEthers.provider.getCode(txInfo.deployData);
+
+    const tx = {
+      from: signer.address,
+      to: commonFactory,
+      value: 0,
+      nonce: await signer.getTransactionCount("latest"),
+      maxPriorityFeePerGas: maxPriorityFeePerGas,
+      maxFeePerGas: baseMaxFeePerGas.add(maxPriorityFeePerGas),
+      data: deployData,
+    };
+
+    await (await signer.sendTransaction(tx)).wait();
+  }
+}
+
+async function deployStaking() {
+  await deployInfraKeyless();
+
+  const cloneFactory: TWStatelessFactory = await hardhatEthers.getContractAt(
+    "TWStatelessFactory",
+    infraContracts.cloneFactory.predictedAddress as string,
+  );
+
+  let stakingInterface = new ethers.utils.Interface(NFTStakeAbi.abi);
+  let encodedFunctionData = stakingInterface.encodeFunctionData("initialize", [
+    "0xdd99b75f095d0c4d5112aCe938e4e6ed962fb024",
+    "",
+    [],
+    "0xEeeeeEeeeEeEeeEeEeEeeEEEeeeeEeeeeeeeEEeE",
+    "0xdd99b75f095d0c4d5112aCe938e4e6ed962fb024",
+    1000,
+    1,
+  ]);
+  const deployProxy = await cloneFactory.deployProxyByImplementation(
+    infraContracts.stake721.predictedAddress,
```

Comment:
so all implementations would be deployed as part of the infra step? 

i think this is nice for us, but for a single user deploying on any EVM might get expensive. something to think about for the full flow.

I think here if you take in the staking abi/bytecode as param here then this can be the generic function that can deploy any prebuilt